parsl/app/futures.py

```diff
@@ -56,8 +72,18 @@ def __init__(self, fut: Future, file_obj: File, tid: int) -> None:
         self._tid = tid
         self.file_obj = file_obj
         self.parent = fut
-
+        if app_fut:
+            self.app_fut = app_fut
+        else:
+            self.app_fut = fut
+        self.data_flow_kernel = dfk
         self.parent.add_done_callback(self.parent_callback)
+        # only capture this if needed
+        if self.data_flow_kernel.file_provenance and self.file_obj.scheme == 'file' and os.path.exists(file_obj.path):
+            file_stat = os.stat(file_obj.path)
```

Comment:
this could be refactored with the code at line 48 https://github.com/Parsl/parsl/pull/3754/files#diff-8ddcd7398ccac61957e74e4e419aafc09215aa3723bcb845503481bcc6edd4c9R48 to have a single provenance update on a DataFuture?

Reply:
I put that in the init so that data about an actual file on disk (as opposed to a true future) would be captured as soon as possible. But I think we could defer that until the parent_callback is called at the end of the app.